### ISSUE TITLE
Scene only @export

### ISSUE THREAD

**Issue by TheColorRed** [godotengine/godot-proposals#7453]
### Describe the project you are working on

A tug of war game

### Describe the problem or limitation you are having in your project

I only want some exports to be modifiable in the scene that they are exported in so that the parent cannot modify them.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add a way to only export variables to the scene that the script is attached to (when attached to the root node). Currently if the script is attached to the root node, parent scenes have the ability to modify the exported variable(s).

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

This would expose the variable to the editor, but only in the scene that the script is attached to and not the parent.

```gdscript
@export_local var my_var: NodePath
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

There isn't one that I am aware of.

### Is there a reason why this should be core and not an add-on in the asset library?

This will affect how the inspector displays exported variables.

**Comment by Mickeon** [comment]
> If this enhancement will not be used often, can it be worked around with a few lines of script?

Yes, actually. It's always possible to override the `_get_property_list` method to get some pretty creative results (assuming that the script contains the `@tool` annotation to run in the editor).

```GDScript
func _get_property_list():
	var properties = []
	
	if is_inside_tree() and get_tree().edited_scene_root == self:
		properties.append({
			name = "my_property",
			type = TYPE_INT,
			usage = PROPERTY_USAGE_DEFAULT,
		})
	
	return properties

```

Notice the `my_property`.

| Scene | Parent Scene |
| ----- | -------| 
| ![image](https://github.com/godotengine/godot-proposals/assets/66727710/6823ec3e-6041-437b-8b20-1cd282b2958a) | ![image](https://github.com/godotengine/godot-proposals/assets/66727710/a7fd7908-e3a1-480a-bc9c-129690cbd2b9) |

**Comment by TheColorRed** [comment]
@Mickeon That isn't working for me.

```
@tool
extends Control

var unit_scene: PackedScene
var tier1: Control
var tier2: Control
var tier3: Control
var race_lbl: Label

func _get_property_list() -> Array[Dictionary]:
	var properties: Array[Dictionary] = []
	var lst = ['unit_scene','tier1','tier2','tier3','race_lbl']

	if is_inside_tree() and get_tree().edited_scene_root == self:
		print('here')
		for itm in lst:
			properties.append({
				name = itm,
				type = typeof(self[itm]),
				usage = PROPERTY_USAGE_DEFAULT
			})

	return properties
```

Either way, I think it would be nice if this feature was built into the editor.

**Comment by TheColorRed** [comment]
Closing the scene and reopening it worked.

**Comment by poohcom1** [comment]
@TheColorRed I made an addon that's similar to this proposal, in case you're interested: https://github.com/poohcom1/godot-private-exports.

**Comment by TheColorRed** [comment]
> @TheColorRed I made an addon that's similar to this proposal, in case you're interested: https://github.com/poohcom1/godot-private-exports.

I cannot get it working, and there is no documentation on how to use it...

**Comment by poohcom1** [comment]
@TheColorRed The addon is pretty simple, so there wasn't much documentation to add. Keep in mind the addon is for the 4.2 dev version, so that might be where your issue is. I don't want to get too off-topic in this thread too much, so if you have any other issues with installing/using the addon, please create an issue on the addon repository.

**Comment by FlooferLand** [comment]
Bump?
It appears more than a few users want this as part of the core engine, including myself, as creating properties that are only used in one singular scene is extremely common _(ex: local references to images or `AudioStream`s)_.
The only addon allowing this has been archived and the other alternative is to use `@export_group` and name the group "Local" which doesn't stop you from accidentally messing up properties and leaves an un-needed group in the inspector that you shouldn't touch

This is a small change that would be greatly beneficial to addon developers and organization in general, as the user shouldn't be able to change or see an addon scene's local resources unless they choose to open up the scene in the case they want to modify it.

**Comment by peterwilson** [comment]
I asked a similar question on Discord and got [a pretty nice answer from the user Outfrost](https://discord.com/channels/1235157165589794909/1260700855401975849/1348761791898845224)

> outfrost: There are "scene-unique names", you can right click the child node and enable scene-unique name, then drag the node into your script and you'll have a get_node ($) shorthand that doesn't change if you move the child node elsewhere in the hierarchy

So for a hierarchy like this, with a Panel that has two children:

![Image](https://github.com/user-attachments/assets/60f60569-402b-4743-9c1f-a4a5863cf6ff)

Instead of `$Panel/TotalValueBar` I can use `%TotalValueBar`

![Image](https://github.com/user-attachments/assets/3c0d02a7-44cd-4689-a774-c04043a957ca)

My original solution was to set local variables in the `_ready` function and use those variables throughout the script. This would still work if you want to avoid littering your script with the `%` shorthand.

![Image](https://github.com/user-attachments/assets/4e40c9d8-9722-4f80-82c6-5e40592e1e90)

**Comment by poohcom1** [comment]
I found myself coming back to this issue a lot. In C# I've been using `GetNode()` or hardcoded resource path, but having an export will make it much more convenient. Furthermore, using GetNode means that the exported node is not available until the node is ready, which limits its use compared to exports. But I still hesitate to use exports with tool scripts because of how values can get reset if the script compilation fails, and silently break scenes.
 
If we want this implement, we'll have to come up with a proper implementation plan, and I have some ideas for that. Note: I'll use 'scene-only' to refer to this feature in general.
 
## Requirements
 
Based on this thread, there are 2 main requirements, and they have different effort for implementation:
 
 1. **Editor hiding**: Hiding scene-only export to prevent clutter.
 2. **Editor validation**: Preventing scene-only exports from being set from parent scene (like `null` values from broken tool scripts or hardcoded value in .tscn files). If the tscn file of the parent scene contains a variable set like `text = "Attack"`, but it belongs to a scene-only export from a child scene, the value should either be ignored or removed. 
 
The former can already be achieved with tool scripts/addon, and would be quite simple to add to core for the same reasons. The latter is a bit more difficult, as it requires extra logic inside the editor. Seems like for most people requirement 1 is more than enough, but I personally also want 2.
 
## Implementation
 
There are 2 parts to implement:
 
 1. A property usage flag for scene-only logic (i.e. `PROPERTY_USAGE_SCENE_ONLY = 2^30`). This is the most logical way to implement something like this, as 'scene-only' is a non-exclusive metadata for exports, similar to categories and groups. Property hint wouldn't make sense because it'll overlap with existing ones (e.g. an export can be both `export_multiline` and `scene_only`).
    - This might be a real blocker because the core maintainers are pretty reluctant to add any new property usages as we're running out of space for a 32-bit int flag. There are some deprecated flags, so we might be able to use those.
 2. An annotation/attribute style for GDScript and C#. This will require some design decisions:
    - What will the annotation be called? How will the annotation apply? To the next var? To every var after until it's disabled (like categories/group)? We don't really have precedence for multiple annotations on variables.
    - How we/should we create a special attribute for C#? Should it stack with the existing one (`[Export][SceneOnly]`), or only require only one attribute like tool button (`[ExportSceneOnly]`).
 
Here's my opinion on these points:
 
For gdscript it should be a stacked annotation. Making an exclusive annotation doesn't make sense as we'll need one for every existing type (i.e. `@export_multiline_sceneonly`, `@export_enum_sceneonly`).
```gdscript
 @export var text: String
 @export var count: int
 @export @scene_only var max_count: int
 @export_multiline @scene_only var default_text: String
 @export var sub_label: Label
```
 
For C#, unlike with GDScript, there's mainly one `[Export]`, and modifications are done through attribute args: `[Export(PropertyHint.Multiline)]`. While you can already make this work with `[Export(propertyUsage: PROPERTY_USAGE_DEFAULT | PROPERTY_USAGE_SCENE_ONLY)]`, I think we should just make a separate exclusive attribute for convenience, similar to `[ExportToolButton]`.
```cs
 [Export] public string Text;
 [Export] public int Count;
 [ExportSceneOnly] public int MaxCount;
 [ExportSceneOnly(PropertyHint.Multiline)] public string DefaultText;
 [Export] public Label SubLabel;
```
 
Internally, all these will do is just add `PROPERTY_USAGE_SCENE_ONLY` to the exported variables.
 Let me know if anyone else has other opinions on the annotation/attribute style.
 
## TL;DR
 
I think if someone wants to implement this, they'll need to get agreement with the core maintainers on the following points:
 
 1. Do we want to implement both editor hiding and editor validation?
 2. Can we expand property usages? Or use deprecated ones?
 3. General annotation/attribute styles for GDScript and C#.
 
I suggest making a proposal first in the Godot contributors chat before starting the PR.
 
I'm fairly interested in working on this, but I'll be quite busy for the next few months, so if someone else is interested, that would be nice.

**Comment by Mickeon** [comment]
I think the most straightforward way to do this is to hide private, yet exported properties, similar to the [previously brought up addon](https://github.com/poohcom1/godot-private-exports). Doing this would make the second goal, "Preventing scene-only exports from being set from parent scene", instinctually possible.

A new property `PROPERTY_USAGE_SCENE_ONLY` usage flag feels particularly excessive for this purpose because it's only useful in the editor. I know that (unfortunately) there are precedents, but there are also alternatives.

**Comment by poohcom1** [comment]
> I think the most straightforward way to do this is to hide private, yet exported properties

This would only apply to C#, no? Or is there something similar in GDScript I'm not aware of?

(altho as a C# user I don't mind having this be a C#-only feature first, but idk if it'll be accepted without language feature parity)

**Comment by denniscarr** [comment]
> I think the most straightforward way to do this is to hide private, yet exported properties, similar to the [previously brought up addon](https://github.com/poohcom1/godot-private-exports). Doing this in core would make the second goal, "Preventing scene-only exports from being set from parent scene", instinctually possible. Of course the user doesn't want the external scenes modifying a private property.
> 
> A new property `PROPERTY_USAGE_SCENE_ONLY` usage flag feels particularly excessive for this purpose because it's only useful in the editor. I know that (unfortunately) there are precedents, but there are also alternatives.

I think it's quite common to mark most exported variables as private in order to make it clear that they should not be modified from other scripts. It would be jarring for many of us if doing this also caused the property to be hidden in the inspector.

That said, I do think this functionality would be useful. I often create an export category called 'internal' to make it clear which properties/node references should not be modified on instances. It would be great to be able to just hide them.
